tests/operator: Fix upgrade test upgrading from both patch and minor release

fossedihelm opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Fix the upgrade test to correctly select the latest release from the
previous minor version (Y-1) instead of falling back to Y-2 during
post-GA windows.

The bug occurred when tests ran on main after a GA release was tagged
on a release branch. The semver comparison logic excluded GA releases
when the tag hint on main was a prerelease version (e.g., v1.7.0-beta.0),
causing tests to skip an entire minor release.
Fixes: https://github.com/kubevirt/kubevirt/issues/16542

Alongside, this introduces testing upgrades from both z-1 and y-1
release.

### References
<!-- optional,
  Use `Fixes #<issue number>(, Fixes #<issue_number>, ...)` format, to close the issue(s) when PR gets merged.
  Use `Partially addresses #<issue number>` to link an issue without closing it when the PR merges.
  
- Fixes #
- Partially addresses #
-->
<!-- optional,
  VEP tracking issue if this PR is implementing one.
  For additional info about VEP tracking issue, see https://github.com/kubevirt/enhancements#process
  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issues/<vep_tracking_issue_number>
-->

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
none
```
/cc @xpivarc @vladikr

fossedihelm committed:
tests/operator: Fix upgrade test to select correct previous minor release

Fix the upgrade test to correctly select the latest release from the
previous minor version (Y-1) instead of falling back to Y-2 during
post-GA windows.

The bug occurred when tests ran on main after a GA release was tagged
on a release branch. The semver comparison logic excluded GA releases
when the tag hint on main was a prerelease version (e.g., v1.7.0-beta.0),
causing tests to skip an entire minor release.
Fixes: https://github.com/kubevirt/kubevirt/issues/16542

Alongside, this also fix a wrong tag selection in the stable branches:
The previous tag selected was the most recent tag, including the z.
This means that the upgrade test were performed from 1.y.n-1 instead of
1.y-1.z.

Signed-off-by: fossedihelm <ffossemo@redhat.com>

akalenyu reviewed (commented):
Just something to consider, when you run inside of prow you should have env vars telling you the base - https://docs.prow.k8s.io/docs/jobs/
`PULL_BASE_REF`

Reference for CDI code https://github.com/kubevirt/containerized-data-importer/commit/219929ee3684a6d2e23facc2fe2a2a33041121a5

fossedihelm commented:
Oh wait! I didn't understand what you mean. Interesting...


Thanks

fossedihelm commented:
/test pull-kubevirt-e2e-k8s-1.34-sig-operator

lyarwood reviewed (commented):
I generated the following slop bug to track this https://github.com/kubevirt/kubevirt/issues/16542 btw

fossedihelm commented:
/test pull-kubevirt-build

lyarwood commented on tests/operator/operator.go:
Could we pull this logic out into a package somewhere and add some basic tests?

lyarwood commented on tests/operator/operator.go:
```diff
@@ -2470,60 +2470,61 @@ func detectLatestUpstreamOfficialTag() (string, error) {
+		// If the targetVersion is the latest release (prerelease excluded) then directly use it
+		if i == 0 && v.Major == targetVersion.Major && v.Minor == targetVersion.Minor {
+			return release.GetTagName(), nil
+		}
```
Ah that fixes the bug I raised around 1.7.0-beta.0 during 1.8 devel on main but I still think this breaks after GA.

fossedihelm committed:
tests/operator: Extract version detection to testable package

Extract version detection logic into a separate `version` package with
comprehensive unit test coverage. This refactoring makes the previous
release detection logic testable and maintainable.

Changes:
- Create new `tests/operator/version` package
- Add DetectLatestUpstreamOfficialTag() function to detect previous
  minor release from GitHub API
- Extract detectLatestUpstreamOfficialTagFromReleases() as testable
  helper that accepts releases and tag hint as parameters
- Add getTagHint() to detect current branch's tag from git

Tests:
- Add 15 unit tests with mocked GitHub releases
- Test GA release scenarios (v1.8.0 -> v1.7.x)
- Test prerelease scenarios (alpha/beta tags)
- Test filtering logic (drafts, prereleases, assets)
- Test major version transitions
- Test error cases and sorting behavior
- Include test for v1.7.0 GA window bug from issue 16542

The refactored code will be used by the operator upgrade tests to
correctly select the previous minor release version.

Related: https://github.com/kubevirt/kubevirt/issues/16542

Assisted-by: Claude <noreply@anthropic.com>
Signed-off-by: fossedihelm <ffossemo@redhat.com>

lyarwood reviewed (commented):
Awesome thanks @fossedihelm! some nits but the tests make this so much easier to review so thanks for that!

fossedihelm commented on tests/operator/version/version_test.go:
```diff
@@ -0,0 +1,243 @@
+	})
+
+	Context("realistic kubevirt release scenarios", func() {
+		It("should handle the v1.7.0 GA window bug scenario", func() {
```
Added the link as comment

lyarwood reviewed (approved):
Thanks @fossedihelm!

xpivarc commented on tests/operator/operator.go:
```diff
@@ -2470,60 +2470,62 @@ func detectLatestUpstreamOfficialTag() (string, error) {
-
-		return err
-	}, 10*time.Second, 1*time.Second).ShouldNot(HaveOccurred())
+	targetTag := getTagHint()
```
Seems like you can return the semver struct?

fossedihelm committed:
tests/operator: Refactor version detection to support Y and Z releases

Split release detection to support both Y (previous minor) and Z
(previous patch) version detection in a single efficient pass.

Assisted-by: Claude <noreply@anthropic.com>
Signed-off-by: fossedihelm <ffossemo@redhat.com>

fossedihelm committed:
tests/operator: perform z-1 and y-1 upgrade tests

Signed-off-by: fossedihelm <ffossemo@redhat.com>

lyarwood commented on tests/operator/version/version.go:
```diff
@@ -0,0 +1,174 @@
+	}
+
+	// Fetch all releases
+	releases, _, err := client.Repositories.ListReleases(context.Background(), "kubevirt", "kubevirt", &github.ListOptions{PerPage: 100})
```
The old code wrapped this call in `Eventually` with a 10-second timeout and 1-second polling interval to handle transient network failures. This is a single-shot call now which could cause flaky test failures in CI. Consider restoring the retry logic.

fossedihelm replied:
done

lyarwood commented on tests/operator/version/version_suite_test.go:
```diff
@@ -0,0 +1,30 @@
+	"kubevirt.io/client-go/testutils"
+)
+
+func TestControllers(t *testing.T) {
```
nit: `TestControllers` should be `TestVersion` to match the package name.

fossedihelm replied:
done

lyarwood commented on tests/operator/operator.go:
```diff
@@ -654,6 +657,32 @@ var _ = Describe("[sig-operator]Operator", Serial, decorators.SigOperator, func(
+				case fromZ:
+					previousImageTag = prevZ
+				}
+				By(fmt.Sprintf("By Using detected tag %s for previous kubevirt", previousImageTag))
```
nit: `By` already marks a step, so this reads as "STEP: By Using detected tag...". Drop the leading "By" from the format string here and on line 683.

fossedihelm replied:
done

lyarwood commented on tests/operator/version/version.go:
```diff
@@ -0,0 +1,174 @@
+}
+
+func getTagHint() (string, error) {
+	//git describe --tags --abbrev=0 "$(git rev-parse HEAD)"
```
nit: missing space after `//` per Go style.

fossedihelm replied:
done

lyarwood reviewed (commented):
@fossedihelm My only concern is the GitHub API call lost its retry logic which could cause CI flakiness. Otherwise I've just got a few minor nits and the PR needs a rebase.

nunnatsa commented on tests/operator/version/version.go:
```diff
@@ -0,0 +1,174 @@
+	"time"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/google/go-github/v32/github"
```
[v32.0.0](https://github.com/google/go-github/releases/tag/v32.0.0) is from May 28, 2020. It maybe even has COVID19!!

It feels risky to me to keep using it against github, that gets updates all the time. Maybe it's the time to bump it?

fossedihelm replied:
done

nunnatsa reviewed (commented):
Hi @fossedihelm 

I added a comment inline.

But my main concern is: will this code work for versions like `v1.9.0-alpha.0`, `v1.9.0-beta.0` or `v1.9.0-rc.0`

fossedihelm committed:
deps: upgrade go-github from v32 to v83

Upgrade the go-github dependency to the latest v83 release, which brings
support for newer GitHub API features. Update all operator test imports
and build files to use the new version. Also removes unused
golang.org/x/crypto packages that are no longer needed.

Signed-off-by: fossedihelm <ffossemo@redhat.com>
Assisted-by: Claude <noreply@anthropic.com>

fossedihelm commented:
> Hi @fossedihelm
> 
> I added a comment inline.
> 
> But my main concern is: will this code work for versions like `v1.9.0-alpha.0`, `v1.9.0-beta.0` or `v1.9.0-rc.0`

@nunnatsa Thanks for the review! I've tested several scenarios, and they all worked; also, there are some unit tests that should cover them.

Updated go-github dependency to v83 (the latest version has go1.25 dependency)

fossedihelm commented:
> @fossedihelm My only concern is the GitHub API call lost its retry logic which could cause CI flakiness. Otherwise I've just got a few minor nits and the PR needs a rebase.

@lyarwood I totally missed your review
Let me address your comments.. apologies

iholder101 commented on tests/operator/version/version_test.go:
```diff
@@ -0,0 +1,244 @@
+				createMockRelease("v1.8.1", notDraft, notPrerelease, 1),
+				createMockRelease("v1.8.0", notDraft, notPrerelease, 1),
+				createMockRelease("v1.7.3", notDraft, notPrerelease, 1),
+				createMockRelease("v1.7.2", notDraft, notPrerelease, 1),
```
nit

Would you mind switching these so they are not sorted? e.g.:
```go
				createMockRelease("v1.8.1", notDraft, notPrerelease, 1),
				createMockRelease("v1.8.2", notDraft, notPrerelease, 1),
				createMockRelease("v1.8.0", notDraft, notPrerelease, 1),
...
```

This will make the test more robust to situations where github doesn't sort versions

iholder101 reviewed (commented):
Thank you very much @fossedihelm!
Looks perfect! See the tiny nit below, feel free to ignore it.

/approve

@lyarwood do you mind having another quick look?

fossedihelm commented:
@iholder101 Thank you!
[fixed](https://github.com/kubevirt/kubevirt/compare/01221446c763e7cb085f54f1c02c40557a7b17cc..e07270088a4d6d54c6fffcaa175b5903cfd692e2)  your comment & [rebase](https://github.com/kubevirt/kubevirt/compare/e07270088a4d6d54c6fffcaa175b5903cfd692e2..b32f7dda36e88c472280bccf318db893d2be0eef)
Thanks

lyarwood reviewed (commented):
/lgtm

fossedihelm commented:
/cherrypick release-1.8 release-1.7 release-1.6
